# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/MediaEventSource.h
# Commit: c34ea4d71235
# Full Hash: c34ea4d71235c0519020893ba04d7206e7fe4f14
# Author: Michael Froman <mfroman@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 1
# Description:
#   Bug 1654112 - fix tsan error, using a destroyed mutex. r=ng
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D121007
# ==============================================================================

diff -r f547e3673690 -r c34ea4d71235 dom/media/MediaEventSource.h
--- a/dom/media/MediaEventSource.h	Wed Jul 21 07:51:58 2021 -0700
+++ b/dom/media/MediaEventSource.h	Mon Jul 26 16:42:40 2021 -0500
@@ -540,6 +540,8 @@
   template <typename T>
   using ArgType = typename detail::EventTypeTraits<T>::ArgType;
 
+  NS_INLINE_DECL_THREADSAFE_REFCOUNTING(MediaEventForwarder);
+
   explicit MediaEventForwarder(nsCOMPtr<nsISerialEventTarget> aEventTarget)
       : mEventTarget(std::move(aEventTarget)) {}
 
@@ -553,15 +555,11 @@
     mListeners = std::move(aOther.mListeners);
   }
 
-  ~MediaEventForwarder() { MOZ_ASSERT(mListeners.IsEmpty()); }
-
   void Forward(MediaEventSource<Es...>& aSource) {
-    // Forwarding a rawptr `this` here is fine, since the dtor disconnects the
-    // listener on the thread used to run the callback (which checks the
-    // disconnect status).
-    mListeners.AppendElement(
-        aSource.Connect(mEventTarget, [this](ArgType<Es>&&... aEvents) {
-          this->NotifyInternal(std::forward<ArgType<Es>...>(aEvents)...);
+    mListeners.AppendElement(aSource.Connect(
+        mEventTarget,
+        [self = RefPtr<MediaEventForwarder>(this)](ArgType<Es>&&... aEvents) {
+          self->NotifyInternal(std::forward<ArgType<Es>...>(aEvents)...);
         }));
   }
 
@@ -573,6 +571,8 @@
   }
 
  private:
+  ~MediaEventForwarder() { MOZ_ASSERT(mListeners.IsEmpty()); }
+
   const nsCOMPtr<nsISerialEventTarget> mEventTarget;
   nsTArray<MediaEventListener> mListeners;
 };